### DISCUSSION TITLE
Model Hash Support in ComfyUI Workflows

### DISCUSSION THREAD

**Discussion by bigcat88** [Comfy-Org/ComfyUI_frontend#1806] [Ideas]
**Good day to everyone!**

There have been previous discussions in the ComfyUI repository about how users could identify where to download models for their workflows. However, aside from using hashes, no practical solution was suggested. @huchenlei mentioned that creating a discussion here would be the best way to move forward.  

I didn’t want to rush this, as I’ve been working on improving the model catalog for Visionatrix to enable one-click model additions while exploring what CivitAI can offer. This process has given me insights into how we can make things more convenient for users and what, from the perspective of a workflow creator and ComfyUI dev-ops/admin, I’d like to see in ComfyUI to simplify providing model information directly in workflows.  

Now, I feel ready to share an idea:  

Since it’s important to support hashes in both regular workflows and API-based workflows, I propose storing hashes within file names as part of the data structure. This approach minimizes changes to the backend and avoids significant structural expansion.  

For example, in the API-based workflow format, we currently have:  

```json
"inputs": {
  "clip_name1": "t5xxl_fp16.safetensors",
  "clip_name2": "clip_l.safetensors",
  "type": "flux"
}
```

And in the regular workflow format:  

```json
"widgets_values": [
  "clip_l.safetensors",
  "t5xxl_fp8_e4m3fn.safetensors",
  "flux"
]
```

What if, instead of just strings, we add optional support for tuples where the second element contains the hash? For example:  

```json
"widgets_values": [
  ["clip_l.safetensors", "0x123ABC121"],
  ["t5xxl_fp8_e4m3fn.safetensors", "0x84643F2"],
  "flux"
]
```

This would allow users to conveniently search for models on platforms like CivitAI and enable tools or automation plugins to automatically download models for workflows with the click of a button.  

On the ComfyUI backend, the required modifications should be minimal. Essentially, we’d need to handle cases where a tuple is used instead of a string:  

```python
if isinstance(x, tuple):
    x = x[0]  # Extract the model name and ignore the hash
```

What does everyone think? Does this idea look good and worth implementing? Are there any blind spots or better alternatives to consider?

**Comment by MalikKillian** [comment]
I think this would be incredibly helpful in faithfully recreating the workflows of others. I was actually just researching the idea of a node to include model hashes into the workflow. I don't know how slow hashing a 4GB+ model might be but it would be great if this was stored/cached somehow by ComfyUI and added automatically.

**Reply by polarathene** [reply]
You need to read the data from disk and process that to verify the checksum. So you'll likely hit a bottleneck on disk I/O on more systems, but if CPU perf/cores is low or RAM is slow, that may also impact it.

Blake3 is known as a pretty performant cryptographic hash that you could use for checksums to verify integrity. They even have a program [b3sum](https://github.com/BLAKE3-team/BLAKE3#the-b3sum-utility) that you could leverage.

There's also [bao](https://github.com/oconnor663/bao) which is an optimized improvement. With that, you'd have more metadata associated, but it's able to verify specific portions of the file so if there was something off you could bail early. After which you could cache it as trusted.

---

The other option I guess is using BitTorrent with models. Services that provide the model files for direct download could provide those as seeds, you'll get similar verification that way.

HuggingFace has git LFS, so you can clone the repo and checkout the specific model file(s) which already has a checksum AFAIK, so that could be used too.
